### PR TITLE
Refactor to use go/analysis

### PR DESCRIPTION
**Pull request by zalenskivolt**
a first go at using go/analysis as suggested in #24

### CODE REVIEW

File: nakedret.go [hunk 1]
```diff
@@ -210,26 +273,52 @@ func (v *returnsVisitor) Visit(node ast.Node) ast.Visitor {
...
+			if err != nil {
+				log.Printf("failed to format named return fix: %s", err)
+			}
+			//v.pass.Reportf(s.Pos(), "%v naked returns on %v line function", funName, fun.funcLength)
+			//v.pass.Reportf(s.Pos(), "naked return in func `%s` with %d lines of code", funName, fun.funcLength)
+			v.pass.Report(analysis.Diagnostic{
+				Pos:     s.Pos(),
+				End:     s.End(),
+				Message: fmt.Sprintf("naked return in func `%s` with %d lines of code", funName, fun.funcLength),
+				SuggestedFixes: []analysis.SuggestedFix{{
```
Comment (zalenskivolt) [inline nakedret.go:291]: SuggestedFixes do not yet get picked up by golangci-lint, but it's planned,
https://github.com/golangci/golangci-lint/issues/1779#issuecomment-1451023670:

> I think that it's the right thing for a linter to support SuggestedFixes (even if SuggestedFixes is still [an experimental API](https://github.com/golang/tools/blob/fbb25cbbd5ff5845639a3035f1fb7ddfd0ac0624/go/analysis/diagnostic.go#L49)), I also believe that we have to remove our custom implementation.

File: cmd/nakedret/main.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+)
+
+const (
+	DefaultLines = 5
+)
+
+func main() {
+	analyzer := nakedret.NakedReturnAnalyzer(DefaultLines)
+	singlechecker.Main(analyzer)
+}
```
Comment (zalenskivolt) [inline cmd/nakedret/main.go:16]: moved `main()` to `cmd/nakedret/main.go` so `nakedret` code can use `package nakedret` and be imported

**Comment by alexkohler** [comment]
I'm seeing a build failure - I think we'd need to bump up the Go version here:
https://github.com/alexkohler/nakedret/blob/c7b0313ed934516ffa928a08892f9f99c2ce8967/.github/workflows/build-test.yml#L12

Mind bumping this up? Happy to merge after that

**Comment by zalenskivolt** [comment]
Bumped to 1.18

**Comment by zalenskivolt** [comment]
> Bumped to 1.18

The build had this error:
> Error: ../../../go/pkg/mod/golang.org/x/tools@v0.8.0/go/gcexportdata/gcexportdata.go:97:9: undefined: io.ReadAll
> note: module requires Go 1.18

but bumping to for example go1.16 also seems to work fine. Bumping to go1.17 adds the `// indirect` lines in go.mod, and bumping to go1.18 cleans up most of go.sum. I guess 1.18 is fine for most projects by now.

File: nakedret.go [hunk 3]
```diff
@@ -23,41 +29,57 @@ func init() {
...
 }
 
 type returnsVisitor struct {
+	pass      *analysis.Pass
 	f         *token.FileSet
 	maxLength uint
-	root      *returnsVisitor
-	found     bool
 
+	stack []funcInfo
```
Comment (alexandear) [inline nakedret.go:72]: Maybe it's better to use`functions` as the slice's name? Because `stack` is unclear.

File: nakedret_test.go [hunk 4]
```diff
@@ -65,3 +82,25 @@ func TestCheckNakedReturns(t *testing.T) {
...
+}
+
+func TestAllFixes(t *testing.T) {
+	wd, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("Failed to get wd: %s", err)
+	}
+
+	testdata := filepath.Join(wd, "testdata")
+	log.Printf("Running tests on %s", testdata)
```
Comment (alexandear) [inline nakedret_test.go:104]: Maybe `t.Log` instead of `log.Printf`?

Btw., do we really need this log here?

File: nakedret.go (hunk 3 above)
Comment (zalenskivolt) [reply nakedret.go:72]: agree, also thought about this.

updated

File: nakedret_test.go (hunk 4 above)
Comment (zalenskivolt) [reply nakedret_test.go]: removed, not really needed

File: nakedret.go [hunk 5]
```diff
@@ -23,41 +29,57 @@ func init() {
...
+		Requires: []*analysis.Analyzer{inspect.Analyzer},
+	}
+	return analyzer
+}
+
+type NakedReturnRunner struct {
+	MaxLength uint
+}
+
+func (n *NakedReturnRunner) run(pass *analysis.Pass) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (zalenskivolt) [reply nakedret.go]: updated to use any

File: nakedret.go [hunk 6]
```diff
@@ -210,26 +263,56 @@ func (v *returnsVisitor) Visit(node ast.Node) ast.Visitor {
...
+		// Push function info to track returns for this function
 		file := v.f.File(node.Pos())
 		length := file.Position(node.End()).Line - file.Position(node.Pos()).Line
-		return &returnsVisitor{
-			f:           v.f,
-			root:        v.root,
-			maxLength:   v.maxLength,
+		if length == 0 {
+			// consider functions that finish on the same line as they start as single line functions, not zero lines!
+			length = 1
```
Comment (zalenskivolt) [inline nakedret.go:307]: added check to handle one-liners

File: README.md [hunk 7]
```diff
@@ -6,7 +6,7 @@ nakedret is a Go static analysis tool to find naked returns in functions greater
 Install Nakedret via go install:
 
 ```cmd
-go install github.com/alexkohler/nakedret@latest
+go install github.com/alexkohler/nakedret/cmd/nakedret@latest
```
Comment (zalenskivolt) [inline README.md:9]: From some local testing, it seems like the command-line tool now has to be installed with
`go install github.com/alexkohler/nakedret/cmd/nakedret` (with cmd/nakedret added)
but this is the same as for example goimports
`go install golang.org/x/tools/cmd/goimports@latest`

I'm not really sure what are best practices for things like this, it looks like many projects use both cmd/x and the root directory. goimports and golangci-lint both use cmd/x as well

Comment (alexkohler) [reply README.md:9]: > From some local testing, it seems like the command-line tool now has to be installed with
go install github.com/alexkohler/nakedret/cmd/nakedret (with cmd/nakedret added)

Hm, would this break folks that are running things like `go install github.com/alexkohler/nakedret@latest` in CI? (not saying that's the best practice ever :sweat_smile:)

> I'm not really sure what are best practices for things like this, it looks like many projects use both cmd/x and the root directory. goimports and golangci-lint both use cmd/x as well

Seems like `cmd/x` might be the most idiomatic thing to do, but I don't really have strong feelings - I'd be fine with `cmd/` or `cmd/x` :+1:

File: nakedret.go [hunk 8]
```diff
@@ -210,26 +273,52 @@ func (v *returnsVisitor) Visit(node ast.Node) ast.Visitor {
...
+		funName := nestedFuncName(v.stack)
+		if fun.reportNaked && len(s.Results) == 0 && push {
+			sFix := nakedReturnFix(s, fun.funcType)
+			b := &bytes.Buffer{}
+			err := printer.Fprint(b, v.f, sFix)
+			if err != nil {
+				log.Printf("failed to format named return fix: %s", err)
+			}
+			//v.pass.Reportf(s.Pos(), "%v naked returns on %v line function", funName, fun.funcLength)
+			//v.pass.Reportf(s.Pos(), "naked return in func `%s` with %d lines of code", funName, fun.funcLength)
```
[the remark this answers is not in the document]
Comment (zalenskivolt) [reply nakedret.go]: they were the original message and the updated message before suggested fixes.

removed

File: README.md (hunk 7 above)
Comment (zalenskivolt) [reply README.md:9]: > Hm, would this break folks that are running things like go install github.com/alexkohler/nakedret@latest in CI? (not saying that's the best practice ever 😅)

that's true, but I guess then you would have the option to either use a specific version if you don't want to update anything, or update the reference?

maybe even update the README to refer to golangci-lint as the primary way to run it? it's just so much easier to all kinds of linting in one place… still good to be able to run each individual linter as a tool of course!

**Comment by alexandear** [comment]
We could also write in `README.md` that it's possible to run `nakedret` as `go vet` tool:

```
go vet -vettool=$(which nakedret) ./...
```

**Comment by zalenskivolt** [comment]
> We could also write in README.md that it's possible to run nakedret as go vet tool:

I pushed a short sentence to the README.

**Comment by zalenskivolt** [comment]
> Happy to merge after that

@alexkohler Feel feel to push any final updates if you're mostly happy! I enabled the allow edits by maintainers 😄

File: README.md (hunk 7 above)
Comment (alexkohler) [reply README.md:9]: Yeah, I think most workflows should be using `golangci-lint`. I think this is a positive change that'll reduce some tech debt over in `golangci-lint`, so even if this ends up breaking some `go install github.com/alexkohler/nakedret@latest` calls, it's probably worth it IMO.

**Comment by alexkohler** [comment]
Thanks again for the PR :smile:

**Pull request merged**